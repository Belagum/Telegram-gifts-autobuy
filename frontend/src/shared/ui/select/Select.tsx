// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 Vova Orig

import type { SelectHTMLAttributes } from "react";
import React from "react";
import clsx from "clsx";
import "./select.css";

export interface Option<T extends string | number> {
  label: string;
  value: T;
}

export interface SelectProps<T extends string | number> extends SelectHTMLAttributes<HTMLSelectElement> {
  options: Option<T>[];
  error?: string;
}

export const Select = React.forwardRef<HTMLSelectElement, SelectProps<string | number>>(
  ({ className, options, error, ...rest }, ref) => {
    return (
      <div className={clsx("ui-select", { "ui-select--error": Boolean(error) })}>
        <select ref={ref} className={clsx("ui-select__field", className)} {...rest}>
          {options.map((option) => (
            <option key={option.value} value={option.value}>
              {option.label}
            </option>
          ))}
        </select>
        {error && <div className="ui-select__error">{error}</div>}
      </div>
    );
  },
);

Select.displayName = "Select";
