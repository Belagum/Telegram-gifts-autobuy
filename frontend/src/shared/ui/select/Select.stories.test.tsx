// SPDX-License-Identifier: Apache-2.0
// Copyright 2025 Vova Orig

import { render } from "@testing-library/react";
import { Select } from "./Select";
import * as stories from "./Select.stories";

describe("Select stories", () => {
  it("renders default select", () => {
    const { container } = render(
      <Select {...(stories.Default.args ?? {})} options={stories.Default.args?.options ?? []} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders error select", () => {
    const { container } = render(
      <Select {...(stories.WithError.args ?? {})} options={stories.WithError.args?.options ?? []} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
