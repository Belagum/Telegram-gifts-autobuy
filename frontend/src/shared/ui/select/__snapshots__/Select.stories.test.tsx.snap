// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select stories > renders default select 1`] = `
<div
  class="ui-select"
>
  <select
    class="ui-select__field"
  />
</div>
`;

exports[`Select stories > renders error select 1`] = `
<div
  class="ui-select ui-select--error"
>
  <select
    class="ui-select__field"
  />
  <div
    class="ui-select__error"
  >
    Выберите значение
  </div>
</div>
`;
